### ISSUE TITLE
Add a way for C/C++ code compiled as a part of a CodeQL test to detect it is being tested

### ISSUE THREAD

**Issue by gsingh93** [github/codeql#9425]
It would be convenient if there was a way for C/C++ code that is being compiled as part of a CodeQL test to know that it's part of a CodeQL test. This could be done by defining some type of macro that is defined when compiling any test, i.e. `__CODEQL_TEST__`.

The reason this would be useful is that I could easily compile and run this code outside of a CodeQL test without making any code modifications. Currently I am doing something like this:
```
#ifndef REAL_COMPILATION

extern int stdin;
extern char* fgets(char*, int, int);
extern char* strchr(const char*, int);
char getchar();
int atoi(char*);
int printf(const char*, ...);
void* malloc(unsigned long);
void puts(const char*);

#else

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#endif
```

And then when I want to test my code outside of CodeQL I do `gcc test.c -DREAL_COMPILATION=1`. This works for me, but it would be more convenient if CodeQL itself defined some macro that I could check, instead of requiring me to specify this when I compile the code manually.

**Comment by tausbn** [comment]
Thank you for your suggestion! I will pass it on to the C++ team.

**Comment by gsingh93** [comment]
Ping 🙂 It seems like this would be a fairly straightforward thing to implement.

**Comment by jketema** [comment]
We've added a `__CODEQL_TEST__` macro. This should be available from CodeQL 2.16.0 onwards.
